src/core/include/units/magnitude.h

```diff
@@ -0,0 +1,360 @@
+ */
+template<typename T>
+concept BasePower = detail::is_base_power<T>;
+
+/**
+ * @brief  Equality detection for two base powers.
+ */
+template<BasePower T, BasePower U>
+constexpr bool operator==(T t, U u) {
+  return std::is_same_v<T, U> && (t.get_base() == u.get_base()) && (t.power == u.power);
+}
+
+/**
+ * @brief  A BasePower, raised to a rational power E.
+ */
+template<ratio E>
+constexpr auto pow(BasePower auto bp) {
+  bp.power = bp.power * E;
+  return bp;
+}
+
+// A variety of implementation detail helpers.
+namespace detail {
+
+// Find the smallest prime factor of `n`.
+constexpr std::intmax_t find_first_factor(std::intmax_t n)
+{
+  for (std::intmax_t f = 2; f * f <= n; f += 1 + (f % 2))
+  {
+    if (n % f == 0) { return f; }
+  }
+  return n;
+}
+
+// The exponent of `factor` in the prime factorization of `n`.
+constexpr std::intmax_t multiplicity(std::intmax_t factor, std::intmax_t n)
+{
+  std::intmax_t m = 0;
+  while (n % factor == 0)
+  {
+    n /= factor;
+    ++m;
+  }
+  return m;
+}
+
+// Divide a number by a given base raised to some power.
+//
+// Undefined unless base > 1, pow >= 0, and (base ^ pow) evenly divides n.
+constexpr std::intmax_t remove_power(std::intmax_t base, std::intmax_t pow, std::intmax_t n)
+{
+  while (pow-- > 0) { n /= base; }
+  return n;
+}
+
+// Helpers to perform prime factorization at compile time.
+template<std::intmax_t N>
+  requires (N > 0)
+struct prime_factorization;
+
+template<std::intmax_t N>
+static constexpr auto prime_factorization_v = prime_factorization<N>::value;
+
+// A way to check whether a number is prime at compile time.
+constexpr bool is_prime(std::intmax_t n) { return (n > 1) && (find_first_factor(n) == n); }
+
+constexpr bool is_valid_base_power(const BasePower auto &bp) {
+  if (bp.power == 0) { return false; }
+
+  if constexpr (std::is_same_v<decltype(bp.get_base()), int>) { return is_prime(bp.get_base()); }
+  else { return bp.get_base() > 0; }
+}
+
+// A function object to apply a predicate to all consecutive pairs of values in a sequence.
+template<typename Predicate>
+struct pairwise_all {
+  Predicate predicate;
+
+  template<typename... Ts>
+  constexpr bool operator()(Ts&&... ts) const {
+    // Carefully handle different sizes, avoiding unsigned integer underflow.
+    constexpr auto num_comparisons = [](auto num_elements) {
+      return (num_elements > 1) ? (num_elements - 1) : 0;
+    }(sizeof...(Ts));
+
+    // Compare zero or more pairs of neighbours as needed.
+    return [this]<std::size_t... Is>(std::tuple<Ts...> &&t, std::index_sequence<Is...>) {
+      return (predicate(std::get<Is>(t), std::get<Is + 1>(t)) && ...);
+    }(std::make_tuple(std::forward<Ts>(ts)...), std::make_index_sequence<num_comparisons>());
+  }
+};
+
+// Deduction guide: permit constructions such as `pairwise_all{std::less{}}`.
+template<typename T>
+pairwise_all(T) -> pairwise_all<T>;
```

Comment:
Well, all of the C++ standard algorithms take the predicate as a function argument rather than a template parameter 😉 so I think it is less surprising to the user than introducing a stateful functor. Also, one function template should be faster to compile than two (class and member function). Although the class template for `std::less{}` will probably be memoized in many cases.

Reply:
Good points.  Although I think the C++ standard algorithms take iterators (or possibly ranges), as opposed to a variadic number of arguments.  I think including the predicate along with a variadic number of arguments feels different, and the separation feels more important.

Regardless, I think this is something worth revisiting in its own PR later if we choose.